modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotRestoreProcess.java

```diff
@@ -1227,6 +1273,379 @@ private void finishCacheStart(UUID reqId, Map<UUID, Boolean> res, Map<UUID, Exce
+
+        UUID incSnpId = ctx.cache().context().snapshotMgr()
+            .readIncrementalSnapshotMetadata(snpName, snpPath, incIdx)
+            .requestId();
+
+        IncrementalSnapshotFinishRecord incSnpFinRec = readFinishRecord(segments[segments.length - 1], incSnpId);
+
+        if (incSnpFinRec == null)
+            throw new IgniteCheckedException("System WAL records for incremental snapshot wasn't found [id=" + incSnpId + ']');
+
+        CacheStripedExecutor exec = new CacheStripedExecutor(ctx.pools().getStripedExecutorService());
+
+        UUID prevIncSnpId = incIdx > 1
+            ? ctx.cache().context().snapshotMgr().readIncrementalSnapshotMetadata(snpName, snpPath, incIdx - 1).requestId()
+            : null;
+
+        long start = U.currentTimeMillis();
+
+        LongAdder applied = new LongAdder();
+
+        try (WALIterator it = walIter(log, segments)) {
+            IgnitePredicate<GridCacheVersion> txVerFilter = null;
+
+            // Skips applying WAL until base snapshot record.
+            while (it.hasNext()) {
+                WALRecord rec = it.next().getValue();
+
+                if (rec.type() == CLUSTER_SNAPSHOT) {
+                    ClusterSnapshotRecord snpRec = (ClusterSnapshotRecord)rec;
+
+                    if (snpRec.clusterSnapshotName().equals(snpName)) {
+                        if (prevIncSnpId == null)
+                            txVerFilter = txVer -> !incSnpFinRec.excluded().contains(txVer);
+
+                        break;
+                    }
+                }
+            }
+
+            // Apply incremental snapshots.
+            while (it.hasNext()) {
+                WALRecord rec = it.next().getValue();
+
+                if (rec.type() == INCREMENTAL_SNAPSHOT_START_RECORD) {
+                    IncrementalSnapshotStartRecord startRec = (IncrementalSnapshotStartRecord)rec;
+
+                    if (startRec.id().equals(incSnpFinRec.id()))
+                        txVerFilter = v -> incSnpFinRec.included().contains(v);
+                }
+                else if (rec.type() == INCREMENTAL_SNAPSHOT_FINISH_RECORD) {
+                    IncrementalSnapshotFinishRecord finRec = (IncrementalSnapshotFinishRecord)rec;
+
+                    if (finRec.id().equals(prevIncSnpId))
+                        txVerFilter = txVer -> !incSnpFinRec.excluded().contains(txVer);
+                }
+                else if (rec.type() == DATA_RECORD_V2) {
+                    DataRecord data = (DataRecord)rec;
+
+                    DataEntry entry = data.writeEntries().get(0);
+
+                    if (txVerFilter == null || txVerFilter.apply(entry.nearXidVersion())) {
+                        for (DataEntry e: data.writeEntries()) {
+                            // That is OK to restore only part of transaction related to a specified cache group,
+                            // because a full snapshot restoring does the same.
+                            if (!cacheIds.contains(e.cacheId()))
```

Comment:
It seems we can move this part of filter inside `txVerFilter` and perform all filtration inside one lambda.

Reply:
`txVerFilter` is `null` while intermediate incremental snapshots applies. For example, user ask to restore incIdx=3. Then Ignite doesn't filter entries from incremental snapshots 1 and 2.